lib/internal/streams/writable.js

```diff
@@ -445,16 +445,17 @@ function _write(stream, chunk, encoding, cb) {
 
   if (typeof encoding === 'function') {
     cb = encoding;
-    encoding = (state[kState] & kDefaultUTF8Encoding) !== 0 ? 'utf8' : state.defaultEncoding;
-  } else {
-    if (!encoding)
-      encoding = (state[kState] & kDefaultUTF8Encoding) !== 0 ? 'utf8' : state.defaultEncoding;
-    else if (encoding !== 'buffer' && !Buffer.isEncoding(encoding))
-      throw new ERR_UNKNOWN_ENCODING(encoding);
-    if (typeof cb !== 'function')
-      cb = nop;
+    encoding = null;
   }
 
+  if (!encoding)
+    encoding = (state[kState] & kDefaultUTF8Encoding) !== 0 ? 'utf8' : state[kDefaultEncodingValue];
+  else if (encoding !== 'buffer' && !Buffer.isEncoding(encoding))
+    throw new ERR_UNKNOWN_ENCODING(encoding);
+
+  if (cb == null || typeof cb !== 'function')
```

Comment:
Nit: I'm pretty sure this is the same as:


```suggestion
  if (typeof cb !== 'function')
```

Because null/undefined are also not functions.

Reply:
it's a micro optimization

Reply:
Although it's micro, the checking typeof cb !== 'function' effectively covers cases where cb is null or undefined, is there any side-effect caused to consider that change as not optimization?

Reply:
Sorry. I don't get the question.

Reply:
I'm asking whether removing the explicit check for cb == null and relying solely on typeof cb !== 'function could potentially have side effects or drawbacks, even though it's considered a micro-optimization.

The suggested change is a micro-optimization, but it's also a simplification that assumes cb should only be a function or null/undefined. If that assumption is correct for that particular use case, then it should be safe. So, let's apply it. If there's any doubt that cb could take other values, then the original version is safer and more explicit.

Reply:
This doesn't change the logic in anyway.

Reply:
I agree that the suggestion should be applied. Why checking it twice when `cb == null`? It would prevent future PRs opened for the same reason.

Reply:
@Mifrill, the micro-optimization @ronag was talking about is the explicit `cb == null` check and not its removal because the assumption is that `cb == null` is more common than something else that is not a function and `cb == null` is faster than `typeof cb !== 'function'`. Anyway `typeof` is already (or used to be) very fast.

Reply:
ah, I see now, thanks @lpinca for the clarification 👍  You got me right, I thought removing of the `cb == null` is a micro-optimization that is not necessary to apply.

> cb == null is faster than typeof cb !== 'function'.

that's the point that I missed. Now it's clear